xml_struct_derive/src/serialize/codegen.rs

```diff
@@ -0,0 +1,588 @@
+fn generate_namespace_attrs_call(namespace_attrs: &[XmlAttribute]) -> TokenStream {
+    if !namespace_attrs.is_empty() {
+        let namespace_attrs: Vec<_> = namespace_attrs
+            .iter()
+            .map(|XmlAttribute { name, value }| quote!((#name, #value)))
+            .collect();
+
+        quote! {
+            .with_attributes([
+                #(#namespace_attrs,)*
+            ])
+        }
+    } else {
+        TokenStream::default()
+    }
+}
+
+/// Generates the calls necessary to serialize struct or enum fields as XML
+/// attributes.
+fn generate_attribute_field_calls(attr_fields: &[Field]) -> TokenStream {
+    if !attr_fields.is_empty() {
+        attr_fields
+            .iter()
+            .map(|field| {
+                let name = field_name_to_string_tokens(field);
+                let accessor = &field.accessor;
+                let ty = &field.ty;
+
+                quote! {
+                    <#ty as ::xml_struct::XmlSerializeAttr>::serialize_as_attribute(&#accessor, &mut start_tag, #name);
+                }
+            })
+            .collect()
+    } else {
+        TokenStream::default()
+    }
+}
+
+/// Generates calls to add a new XML element to the document, including any
+/// necessary attributes.
+///
+/// If `content_calls` is `None`, the XML element will be an empty tag (e.g.,
+/// "<SomeTag/>"). Otherwise, the XML element will enclose any content added to
+/// the writer by those calls.
+fn generate_xml_tag_calls(
+    name_tokens: TokenStream,
+    namespace_attrs: &[XmlAttribute],
+    attr_fields: &[Field],
+    content_calls: Option<TokenStream>,
+) -> TokenStream {
+    let namespaces_call = generate_namespace_attrs_call(namespace_attrs);
+    let field_calls = generate_attribute_field_calls(attr_fields);
+
+    let calls = if let Some(content_calls) = content_calls {
+        // If the type has fields to serialize as child elements, wrap them
+        // first in an appropriate parent element.
+        quote! {
+            writer.write_event(
+                ::quick_xml::events::Event::Start(start_tag)
+            )?;
+
+            #content_calls
+
+            writer.write_event(
+                ::quick_xml::events::Event::End(
+                    ::quick_xml::events::BytesEnd::new(#name_tokens)
+                )
+            )?;
+        }
+    } else {
+        // If the type has no fields which are to be serialized as child
+        // elements, write an empty XML tag.
+        quote! {
+            writer.write_event(
+                ::quick_xml::events::Event::Empty(start_tag)
+            )?;
+        }
+    };
+
+    quote! {
+        let mut start_tag = ::quick_xml::events::BytesStart::new(#name_tokens)
+            #namespaces_call;
+
+        #field_calls
+
+        #calls
+    }
+}
+
+/// Generates a series of calls for serializing the given fields as XML content
+/// nodes.
+fn generate_child_node_calls_for_fields(child_fields: Vec<Field>) -> TokenStream {
```

Comment:
This is a place where I struggle a bit with naming, yeah. We need some way to distinguish between serializing a structure as an element or as a child of that element. `child_node` is the best name I've come up with for the latter, and so this is generating calls to `serialize_child_nodes()`, but it's a horrible name.

Is `generate_field_serialization_calls()` any better?

Reply:
Wouldn't `generate_content_node_calls()` work better? Since as far as I can tell this is what the function does, when looked at in isolation, and matches with the documentation comment.